lib_eio/core/fiber.ml

```diff
@@ -284,3 +286,108 @@ let with_binding var value fn =
 let without_binding var fn =
   let ctx = Effect.perform Cancel.Get_context in
   Cancel.Fiber_context.with_vars ctx (Hmap.rem var ctx.vars) fn
+
+(* Coroutines.
+
+   [fork_coroutine ~sw fn] creates a new fiber for [fn]. [fn] immediately suspends, setting its state to
+   [Ready enqueue]. A consumer can resume it by setting the state to [Running] and calling [enqueue],
+   while suspending itself. The consumer passes in its own [enqueue] function. They run alternatively
+   like this, switching between the [Ready] and [Running] states.
+
+   To finish, the coroutine fiber can set the state to [Finished] or [Failed],
+   or the client can set the state to [Client_cancelled].
+*)
+
+(* Note: we could easily generalise this to [('in, 'out) coroutine] if that was useful. *)
+type 'out coroutine =
+  [ `Init
```

Comment:
Wouldn't this be slightly more efficient without using poly variants and just normal ones? Not using subtyping anywhere I think.

Reply:
Ah yes, I'd missed that case of `enqueue`; thanks!

Minor nit: iirc there's a modest efficiency benefit in pattern matching with normal variants as well, since the tags are more predictable. E.g.

```
type t =
| Alice | Bob | Charlie | Dan

let fn1 f = function
  | Alice -> 1
  | Bob -> 2
  | Charlie -> 3
  | Dan -> 4

let fn2 f = function
  |`Alice -> 1
  |`Bob -> 2
  |`Charlie -> 3
  |`Dan -> 4
```
gives

```
% ocamlc -dlambda t.ml
(setglobal T!
  (let
    (fn1/272 =
       (function f/274 param/275[int] : int
         (switch* param/275
          case int 0: 1
          case int 1: 2
          case int 2: 3
          case int 3: 4))
     fn2/276 =
       (function f/278 param/279[int] : int
         (if (>= param/279 482771474) (if (>= param/279 884917024) 1 3)
           (if (>= param/279 3403313) 4 2))))
    (makeblock 0 fn1/272 fn2/276)))
```

Reply:
```ocaml
[@@@warning "-A"]

type t =
| Alice | Bob | Charlie | Dan

let fn1 = function
  | Alice -> 1
  | Bob -> 2
  | Charlie -> 3
  | Dan -> 4

let fn2 = function
  |`Alice -> 1
  |`Bob -> 2
  |`Charlie -> 3
  |`Dan -> 4

let fn1 = Sys.opaque_identity fn1
let fn2 = Sys.opaque_identity fn2

let time fn =
  let t0 = Unix.gettimeofday () in
  for i = 1 to 100_000 do
    fn ()
  done;
  let t1 = Unix.gettimeofday () in
  Printf.printf "%.3f ms\n" (1000. *. (t1 -. t0))

let () =
  print_endline "";
  time (fun () -> ignore (fn1 Alice));
  time (fun () -> ignore (fn2 `Alice));
  time (fun () -> ignore (fn1 Alice));
  time (fun () -> ignore (fn2 `Alice));
```
Any difference seems like noise:
```
0.359 ms
0.349 ms
0.353 ms
0.345 ms
```

Reply:
Indeed, you won't notice it much with small type definitions -- it's only an issue when the pattern matching code in the compiler can't generate an efficient jump table for a large pattern match. (I've run into this when using OCaml code generators in the past, when it's convenient for the generator to just emit polymorphic variants, but going the extra mile for normal variants makes a difference).